db/map.go

```diff
@@ -57,31 +42,19 @@ func (e ErrDatabaseOp) Unwrap() error {
 	return e.Err
 }
 
-// IsNoRows is a utility function for casting an error to ErrDatabaseOp and
-// returning true if its wrapped err is sql.ErrNoRows. If the error is not an
-// ErrDatabaseOp the return value of IsNoRows will always be false.
+// IsNoRows is a utility function for determining if an error wraps the go sql
+// package's ErrNoRows, which is returned when a Scan operation has no more
+// results to return, and as such is returned by many gorp methods.
 func IsNoRows(err error) bool {
-	// if the err is an ErrDatabaseOp instance, return its noRows() result to see
-	// if the inner err is sql.ErrNoRows
-	var dbErr ErrDatabaseOp
-	if errors.As(err, &dbErr) {
-		return dbErr.noRows()
-	}
-	return false
+	return errors.Is(err, sql.ErrNoRows)
 }
 
-// IsDuplicate is a utility function for casting an error to ErrDatabaseOp and
-// returning a boolean indicating if it is a duplicate error or not. If the
-// error is not an ErrDatabaseOp the return value of IsDuplicate will always be
-// false.
+// IsDuplicate is a utility function for determining if an error wrap MySQL's
+// Error 1062: Duplicate entry. This error is returned when inserting a row
+// would violate a unique key constraint.
 func IsDuplicate(err error) bool {
-	// if the err is an ErrDatabaseOp instance, return its duplicate() result to
-	// see if the inner err indicates a duplicate row error.
-	var dbErr ErrDatabaseOp
-	if errors.As(err, &dbErr) {
-		return dbErr.duplicate()
-	}
-	return false
+	var dbErr *mysql.MySQLError
+	return errors.As(err, &dbErr) && dbErr.Number == 1062
```

Comment:
That's an interesting way to have a conditional return.

Reply:
@pgporada I'm glad you noticed this! While this return is rather tame I've seen a perpetrated more than my fair share of one-liner code golfing. My fave was this simple pig latin challenge I wrote in Python:
```python
import string
def pig_it(text):
    return ' '.join([f"{x[1:]}{x[0]}ay" if x not in string.punctuation else x for x in text.split()])
```